bokeh/events.py

```diff
@@ -0,0 +1,178 @@
+''' Represent granular events that can be used to trigger callbacks.
+
+Bokeh documents and applications are capable of supporting various kinds of
+interactions. These are often associated with events, such as mouse or touch
+events, interactive downsampling mode activation, widget or tool events, and
+others. The classes in this module represent these different events, so that
+callbacks can be attached and executed when they occur.
+
+It is possible to respond to events with ``CustomJS`` callbacks, which will
+function with or without a Bokeh server. This can be accomplished with the
+:func:`~bokeh.model.Model.js_on_event` method:
+
+.. code-block:: python
+
+    # example setting up JS callback
+
+Alternatively it is possible to trigger Python code to run when events
+happen, in the context of a Bokeh application running on a Bokeh server.
+This can ccomplished with the :func:`~bokeh.model.Model.on_event` method:
+
+. code-block:: python
+
+    # example setting up python callback
+
+'''
+from __future__ import absolute_import
+
+class Event(object):
+    ''' Base class for all Bokeh events.
+
+    '''
+    event_classes = []
+    event_name = None
+
+    def __init__(self, model_id=None):
+        self.model_id = model_id
+
+    @classmethod
+    def register_event_class(cls, event_cls):
+        ''' Register a custom event class.
+
+        '''
+        cls.event_classes.append(event_cls)
+
+    @classmethod
+    def from_JSON(cls, json):
+        eventclasses = [eventcls for eventcls in cls.event_classes]
+        for eventscls in eventclasses:
+            if eventscls.event_name == json['event_name']:
+                return eventscls(**json['event_values'])
+        print('Warning: Could not find appropriate Event class')
```

Comment:
Instead of this, we should define whatever hook we need to define in order to plug into the standard `json` module so that the usual `json.loads(json_str)` works